File: AdventOfCode2024/Day04/day04.mjs
import * as utils from "../utils.mjs";
import { dirname } from "path";
import { fileURLToPath } from "url";
const scriptDirectory = dirname(fileURLToPath(import.meta.url));
const exampleInput = utils.readInput(scriptDirectory, "example.txt");
const puzzleInput = utils.readInput(scriptDirectory, "input.txt");

function getWord(input, i, j, di, dj) {
  let word = "";
  for (let t = 0; t < 4; t++) {
    let row = input[i + t * di];
    if (!row) {
      return "";
    }
    let letter = input[i + t * di][j + t * dj];
    if (!letter) {
      return "";
    }

    word += letter;
  }

  return word;
}

function countXmas(input, i, j) {
  let cellCount = 0;
  let row = getWord(input, i, j, 0, 1);
  if (row == "XMAS" || row == "SAMX") cellCount++;

  let column = getWord(input, i, j, 1, 0);
  if (column == "XMAS" || column == "SAMX") cellCount++;

  let diag = getWord(input, i, j, 1, 1);
  if (diag == "XMAS" || diag == "SAMX") cellCount++;

  let diag2 = getWord(input, i, j, 1, -1);
  if (diag2 == "XMAS" || diag2 == "SAMX") cellCount++;

  // console.log(i, j, cellCount);

  return cellCount;
}

function countXmas2(input, i, j) {
  if (
    input[i][j] == "A" &&
    ((input[i - 1][j - 1] == "M" &&
      input[i + 1][j - 1] == "M" &&
      input[i - 1][j + 1] == "S" &&
      input[i + 1][j + 1] == "S") ||
      (input[i - 1][j - 1] == "S" &&
        input[i + 1][j - 1] == "S" &&
        input[i - 1][j + 1] == "M" &&
        input[i + 1][j + 1] == "M") ||
      (input[i - 1][j - 1] == "M" &&
        input[i + 1][j - 1] == "S" &&
        input[i - 1][j + 1] == "M" &&
        input[i + 1][j + 1] == "S") ||
      (input[i - 1][j - 1] == "S" &&
        input[i + 1][j - 1] == "M" &&
        input[i - 1][j + 1] == "S" &&
        input[i + 1][j + 1] == "M"))
  ) {
    return 1;
  }
  return 0;
}

function solve(input) {
  // Part 1
  let p1 = 0;
  for (let i = 0; i < input.length; i++) {
    for (let j = 0; j < input[0].length; j++) {
      p1 += countXmas(input, i, j);
    }
  }

  // Part 2
  let p2 = 0;
  for (let i = 1; i < input.length - 1; i++) {
    for (let j = 1; j < input[0].length - 1; j++) {
      p2 += countXmas2(input, i, j);
    }
  }

  return [p1, p2];
}

//27842056
console.log("example:", solve(exampleInput));
console.log(" puzzle:", solve(puzzleInput));

File: AdventOfCode2024/Day04/day04.py
import re

def solve(file):  
  data = open(file).readlines()
  g = { (x, y): c
       for y, r in enumerate(data)
       for x, c in enumerate(r.strip())       
  }
  xh, yh = max(g.keys())
  p1 = sum("XMAS" == "".join(g.get((x + dx * n, y + dy * n), "")
                             for n in range(4))
           for y in range(yh + 1)
           for x in range(xh + 1)
           for dx in (-1, 0, 1)
           for dy in (-1, 0, 1))

  xmas = ( "MAS", "SAM" )
  p2 = sum("".join(g.get((x + dx, y + dy), "")
                   for dx, dy in [(-1,-1), (0,0), (1,1)]) in xmas and
           "".join(g.get((x + dx, y + dy), "")
                   for dx, dy in [(-1,1), (0,0), (1,-1)]) in xmas
           for y in range(yh + 1)
           for x in range(xh + 1))
  return p1, p2

print(solve('example.txt'))
print(solve('input.txt'))